File: Chloe Gwon/2022-06-16-3Sum.js
var threeSum = function(nums) {
    const result = [];
    nums.sort((a,b) => a-b);

    for (var i=0; i<nums.length-2; i++){
        if (i>0 && (nums[i] === nums[i-1])) continue;
        var left = i+1;
        var right = nums.length-1;
    
        while (left < right){
            var threeSum = nums[i] + nums[left] + nums[right];
            if (threeSum > 0){
                right--;
            } else if (threeSum < 0){
                left++;
            } else{
                result.push([nums[i], nums[left], nums[right]]);
                
                while(nums[left] === nums[left+1]){
                    left++;
                }
                // while(nums[right] === nums[right-1]){
                //     right--;
                // }
                left++;
                // right--;
            }
        }
    }
    
    return result;
};

File: Jonghyeok Kim/2022-06-16-3-sum.py
from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        for idx, num in enumerate(nums):
            if idx == 0 or (idx > 0 and nums[idx] <= 0 and nums[idx-1] != nums[idx]):
                left, right = idx + 1, len(nums) - 1
                while left < right:
                    three_sum = num + nums[left] + nums[right]
                    if three_sum > 0:
                        right -= 1
                    elif three_sum < 0:
                        left += 1
                    else:
                        res.append([num, nums[left], nums[right]])
                        left += 1
                        while nums[left] == nums[left-1] and left < right:
                            left += 1
        return res